### PR TITLE
feat: enable user to edit their preferred language and the country they live in 

### LINKED ISSUE #8748: Enable user to edit the preferred communication language field
Currently we get language field from the "inital_lc" value which is determined by the domain name of the subscription form.

Add a user_lc field (preferred communication language).

Let user modify it at subscription time or in their user profile.

On edition:
* if we have a initial_lc, use it to populate user_lc
* else fallback to actual $lc

On subscription:
* prefill with $lc

### Part of
- #5730

### PR DESCRIPTION
**Pull request by albatrousse**
### What

The goal is to add - when signing up and in the user profile - a field for the user to inform their preferred language and the country they live. When signing up, the fields will be prefill with the country associated with the website (looking at the country code) the user uses and the same goes for the language.
For instance, this could lead to better cohesion between user's preferred language and its newsletter 

### Related issue(s) and discussion
- Fixes #8748 and #8749

### CODE REVIEW

**Comment by albatrousse** [comment]
I have created two new sub functions get_countries and get_languages which respectively returns a list of the languages (or the countries) in the language stored in $lc

File: lib/ProductOpener/Display.pm [hunk 1]
```diff
@@ -11430,4 +11430,51 @@ sub generate_select2_options_for_taxonomy_to_json ($target_lc, $tagtype) {
...
+sub get_languages{
+	my @languages_list = ();
+	my @tags_list = get_all_taxonomy_entries("languages");
+	my $language;
+	foreach my $tag (@tags_list){
+		$language = display_taxonomy_tag($lc, "languages", $tag);
+		push @languages_list, $language;
+	}
+	return @languages_list;
+}
```
Comment (alexgarel) [inline lib/ProductOpener/Display.pm:11454]: This would be could to cache it… let's do it at the end :-)

Comment (alexgarel) [reply lib/ProductOpener/Display.pm:11454]: at the end you should sort the languages_list.

Comment (albatrousse) [reply lib/ProductOpener/Display.pm]: I sorted it before returning the list, but I did not quite understand what should I do.
Why caching it is a bad idea ?

**Comment by albatrousse** [comment]
I'm currently trying to correct the functions get_languages() and get_countries() in Display.pm, they are causing errors I don't know why but I will try to resolve it

**Comment by alexgarel** [comment]
@stephanegigandet I added a commit to try to finish this PR. Can you review ?

File: lib/ProductOpener/Web.pm [hunk 2]
```diff
@@ -291,14 +298,21 @@ The data structured can be passed to HTML templates to construction a list of op
 
 sub get_languages_options_list ($target_lc) {
 
+	return $lang_options_cache{$target_lc} if (defined $lang_options_cache{$target_lc});
+
 	my @lang_options = ();
 
 	my %lang_labels = ();
 	foreach my $l (@Langs) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Web.pm:306]: It wasn't changed in this PR, but can we rename $l to $language_code or something? In the GitHub font I can't really tell $l from $1 : `$l / $1`

File: lib/ProductOpener/Web.pm [hunk 3]
```diff
@@ -291,14 +298,21 @@ The data structured can be passed to HTML templates to construction a list of op
...
-		$lang_labels{$l} = display_taxonomy_tag($target_lc, 'languages', $language_codes{$l});
+		my $label = display_taxonomy_tag($target_lc, 'languages', $language_codes{$l});
+		# remove eventual language prefix
+		if ($label =~ /^..:/) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Web.pm:309]: I would replace the whole condition by just `$label =~ s/^\w\w://;`

File: lib/ProductOpener/Web.pm [hunk 4]
```diff
@@ -311,7 +325,53 @@ sub get_languages_options_list ($target_lc) {
...
+
 	return \@lang_options;
 }
 
+# cache for get_countries
+my %countries_options_lists = ();
+
+=head2 get_countries_options_list( $target_lc )
+
+Generates all the countries name in the $lc language suitable for an select option list
```
Comment (stephanegigandet) [inline lib/ProductOpener/Web.pm:339]:
```suggestion
Generates all the countries name in the $target_lc language suitable for an select option list
```

File: lib/ProductOpener/Web.pm [hunk 5]
```diff
@@ -311,7 +325,53 @@ sub get_languages_options_list ($target_lc) {
...
+=head4 $target_lc - language code for labels
+
+=head3 Return value
+
+A reference to a list of hashes with every country code and their label in the $lc language
+[{value => "fr", label => "France"},…]
+
+=cut
+
+sub get_countries_options_list ($target_lc) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Web.pm:352]: There's similar code in /cgi/country.pl which is used to display the drop down of countries in the header of each page. In fact we could reuse the dropdown.

Maybe we could merge the two. The one in country.pl is missing caching, that's a good idea.

File: lib/ProductOpener/Web.pm [hunk 6]
```diff
@@ -311,7 +325,53 @@ sub get_languages_options_list ($target_lc) {
...
+		return $countries_options_lists{$target_lc};
+	}
+	# compute countries list
+	my @countries_list = ();
+	my @tags_list = get_all_taxonomy_entries("countries");
+	foreach my $tag (@tags_list) {
+		next if $tag eq 'en:world';
+		my $country = display_taxonomy_tag($target_lc, "countries", $tag);
+		# remove eventual language prefix
+		if ($country =~ /^..:/) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Web.pm:364]: $country =~ s/^\w\w://;

**Review by stephanegigandet — APPROVED** [review]
Looks good, some minor suggestions. Thank you!

**Pull request merged**